clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
+                DEFAULT_RETRY_BACKOFF_MAX_MS,
+                DEFAULT_HEARTBEAT_JITTER_MS));
 
-    private void setUp(Optional<ConsumerTestBuilder.GroupInformation> groupInfo) {
-        testBuilder = new ConsumerTestBuilder(groupInfo, true, false);
-        time = testBuilder.time;
-        coordinatorRequestManager = testBuilder.coordinatorRequestManager.orElseThrow(IllegalStateException::new);
-        heartbeatRequestManager = testBuilder.heartbeatRequestManager.orElseThrow(IllegalStateException::new);
-        heartbeatRequestState = testBuilder.heartbeatRequestState.orElseThrow(IllegalStateException::new);
-        heartbeatState = testBuilder.heartbeatState.orElseThrow(IllegalStateException::new);
-        backgroundEventHandler = testBuilder.backgroundEventHandler;
-        subscriptions = testBuilder.subscriptions;
-        membershipManager = testBuilder.membershipManager.orElseThrow(IllegalStateException::new);
-        metadata = testBuilder.metadata;
-        metrics = new Metrics(time);
+        this.heartbeatRequestManager = new HeartbeatRequestManager(
+                logContext,
+                pollTimer,
+                config,
+                coordinatorRequestManager,
+                membershipManager,
+                heartbeatState,
+                heartbeatRequestState,
+                backgroundEventHandler,
+                metrics);
 
         when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
+        Map<Uuid, SortedSet<Integer>> map = new HashMap<>();
+        LocalAssignment local = new LocalAssignment(0, map);
+        when(membershipManager.currentAssignment()).thenReturn(local);
     }
 
-    private void resetWithZeroHeartbeatInterval(Optional<String> groupInstanceId) {
-        cleanup();
-
-        ConsumerTestBuilder.GroupInformation gi = new ConsumerTestBuilder.GroupInformation(
-                DEFAULT_GROUP_ID,
-                groupInstanceId,
+    private void createHeartbeatStateWith0HeartbeatInterval() {
+        this.heartbeatRequestState = spy(new HeartbeatRequestState(
+                logContext,
+                time,
                 0,
-                0.0,
-                Optional.of(DEFAULT_REMOTE_ASSIGNOR)
-        );
-
-        setUp(Optional.of(gi));
-    }
+                DEFAULT_RETRY_BACKOFF_MS,
+                DEFAULT_RETRY_BACKOFF_MAX_MS,
+                DEFAULT_HEARTBEAT_JITTER_MS));
 
-    @AfterEach
-    public void cleanup() {
-        if (testBuilder != null) {
-            testBuilder.close();
-        }
+        heartbeatRequestManager = createHeartbeatRequestManager(
+                coordinatorRequestManager,
+                membershipManager,
+                heartbeatState,
+                heartbeatRequestState,
+                backgroundEventHandler);
     }
 
     @Test
     public void testHeartbeatOnStartup() {
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```

Comment:
Not really hurtful but this is changing the path the test runs. Before this PR, the first poll on this test would return null because the member was unsubscribed (shouldSkipHeartbeat true). Now it's because there is no coordinator. Both go down the same path in the poll implementation but if we want to be true to what we had before, here we should mock there is a known coordinator, and when(membershipMgr.shouldSkipHB).thenReturn(true).

Reply:
I have made the change you suggested, let me know what you think.